File: scalardb-sql-jdbc-sample/src/main/java/sample/Sample.java
package sample;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class Sample {

  public void loadInitialData() throws SQLException {
    try (Connection connection = getConnection()) {
      try {
        loadCustomerIfNotExists(connection, 1, "Yamada Taro", 10000, 0);
        loadCustomerIfNotExists(connection, 2, "Yamada Hanako", 10000, 0);
        loadCustomerIfNotExists(connection, 3, "Suzuki Ichiro", 10000, 0);
        loadItemIfNotExists(connection, 1, "Apple", 1000);
        loadItemIfNotExists(connection, 2, "Orange", 2000);
        loadItemIfNotExists(connection, 3, "Grape", 2500);
        loadItemIfNotExists(connection, 4, "Mango", 5000);
        loadItemIfNotExists(connection, 5, "Melon", 3000);

        // Commit the transaction
        connection.commit();
      } catch (Exception e) {
        connection.rollback();
        throw e;
      }
    }
  }

  private void loadCustomerIfNotExists(
      Connection connection, int customerId, String name, int creditLimit, int creditTotal)
      throws SQLException {
    try (PreparedStatement preparedStatementForSelect =
            connection.prepareStatement("SELECT * FROM sample.customers WHERE customer_id = ?");
        PreparedStatement preparedStatementForInsert =
            connection.prepareStatement(
                "INSERT INTO sample.customers (customer_id, name, credit_limit, credit_total) VALUES (?, ?, ?, ?)")) {

      preparedStatementForSelect.setInt(1, customerId);
      try (ResultSet resultSet = preparedStatementForSelect.executeQuery()) {
        if (resultSet.next()) {
          // If the customer info for the specified customer ID already exists, do nothing
          return;
        }
      }

      preparedStatementForInsert.setInt(1, customerId);
      preparedStatementForInsert.setString(2, name);
      preparedStatementForInsert.setInt(3, creditLimit);
      preparedStatementForInsert.setInt(4, creditTotal);
      preparedStatementForInsert.executeUpdate();
    }
  }

  private void loadItemIfNotExists(Connection connection, int itemId, String name, int price)
      throws SQLException {
    try (PreparedStatement preparedStatementForSelect =
            connection.prepareStatement("SELECT * FROM sample.items WHERE item_id = ?");
        PreparedStatement preparedStatementForInsert =
            connection.prepareStatement(
                "INSERT INTO sample.items (item_id, name, price) VALUES (?, ?, ?)")) {

      preparedStatementForSelect.setInt(1, itemId);
      try (ResultSet resultSet = preparedStatementForSelect.executeQuery()) {
        if (resultSet.next()) {
          // If the item info for the specified item ID already exists, do nothing
          return;
        }
      }

      preparedStatementForInsert.setInt(1, itemId);
      preparedStatementForInsert.setString(2, name);
      preparedStatementForInsert.setInt(3, price);
      preparedStatementForInsert.executeUpdate();
    }
  }

  public String getCustomerInfo(int customerId) throws SQLException {
    try (Connection connection = getConnection()) {
      try (PreparedStatement preparedStatement =
          connection.prepareStatement("SELECT * FROM sample.customers WHERE customer_id = ?")) {
        preparedStatement.setInt(1, customerId);

        String name;
        int creditLimit;
        int creditTotal;
        try (ResultSet resultSet = preparedStatement.executeQuery()) {
          if (!resultSet.next()) {
            // If the customer info the specified customer ID doesn't exist, throw an exception
            throw new RuntimeException("Customer not found");
          }

          name = resultSet.getString("name");
          creditLimit = resultSet.getInt("credit_limit");
          creditTotal = resultSet.getInt("credit_total");
        }

        // Commit the transaction
        connection.commit();

        // Return the customer info as a JSON format
        return String.format(
            "{\"id\": %d, \"name\": \"%s\", \"credit_limit\": %d, \"credit_total\": %d}",
            customerId, name, creditLimit, creditTotal);
      } catch (Exception e) {
        connection.rollback();
        throw e;
      }
    }
  }

  public String placeOrder(int customerId, int[] itemIds, int[] itemCounts) throws SQLException {
    assert itemIds.length == itemCounts.length;

    try (Connection connection = getConnection()) {
      try {
        String orderId = UUID.randomUUID().toString();

        // Put the order info into the orders table
        try (PreparedStatement preparedStatement =
            connection.prepareStatement(
                "INSERT INTO sample.orders (customer_id, order_id, timestamp) VALUES (?, ?, ?)")) {
          preparedStatement.setInt(1, customerId);
          preparedStatement.setString(2, orderId);
          preparedStatement.setLong(3, System.currentTimeMillis());
          preparedStatement.executeUpdate();
        }

        int amount = 0;
        for (int i = 0; i < itemIds.length; i++) {
          int itemId = itemIds[i];
          int count = itemCounts[i];

          // Put the order statement into the statements table
          try (PreparedStatement preparedStatement =
              connection.prepareStatement(
                  "INSERT INTO sample.statements (order_id, item_id, count) VALUES (?, ?, ?)")) {
            preparedStatement.setString(1, orderId);
            preparedStatement.setInt(2, itemId);
            preparedStatement.setInt(3, count);
            preparedStatement.executeUpdate();
          }

          // Retrieve the item info from the items table
          try (PreparedStatement preparedStatement =
              connection.prepareStatement("SELECT * FROM sample.items WHERE item_id = ?")) {
            preparedStatement.setInt(1, itemId);

            try (ResultSet resultSet = preparedStatement.executeQuery()) {
              if (!resultSet.next()) {
                // If the item info for the specified item ID doesn't exist, throw an exception
                throw new RuntimeException("Item not found");
              }

              // Calculate the total amount
              amount += resultSet.getInt("price") * count;
            }
          }
        }

        // Check if the credit total exceeds the credit limit after payment
        int updatedCreditTotal;
        try (PreparedStatement preparedStatement =
            connection.prepareStatement("SELECT * FROM sample.customers WHERE customer_id = ?")) {
          preparedStatement.setInt(1, customerId);

          try (ResultSet resultSet = preparedStatement.executeQuery()) {
            if (!resultSet.next()) {
              // If the customer info for the specified customer ID doesn't exist, throw an
              // exception
              throw new RuntimeException("Customer not found");
            }

            int creditLimit = resultSet.getInt("credit_limit");
            int creditTotal = resultSet.getInt("credit_total");
            updatedCreditTotal = creditTotal + amount;

            if (updatedCreditTotal > creditLimit) {
              throw new RuntimeException("Credit limit exceeded");
            }
          }
        }

        // Update credit_total for the customer
        try (PreparedStatement preparedStatement =
            connection.prepareStatement(
                "UPDATE sample.customers SET credit_total = ? WHERE customer_id = ?")) {
          preparedStatement.setInt(1, updatedCreditTotal);
          preparedStatement.setInt(2, customerId);
          preparedStatement.executeUpdate();
        }

        // Commit the transaction
        connection.commit();

        // Return the order id
        return String.format("{\"order_id\": \"%s\"}", orderId);
      } catch (Exception e) {
        connection.rollback();
        throw e;
      }
    }
  }

  private String getOrderJson(Connection connection, String orderId) throws SQLException {
    // Retrieve the order info for the order ID from the orders table
    int customerId;
    long timestamp;
    try (PreparedStatement preparedStatement =
        connection.prepareStatement("SELECT * FROM sample.orders WHERE order_id = ?")) {
      preparedStatement.setString(1, orderId);

      try (ResultSet resultSet = preparedStatement.executeQuery()) {
        if (!resultSet.next()) {
          // If the order info for the specified order ID doesn't exist, throw an exception
          throw new RuntimeException("Order not found");
        }

        customerId = resultSet.getInt("customer_id");
        timestamp = resultSet.getLong("timestamp");
      }
    }

    // Retrieve the customer info for the specified customer ID from the customers table
    String customerName;
    try (PreparedStatement preparedStatement =
        connection.prepareStatement("SELECT * FROM sample.customers WHERE customer_id = ?")) {
      preparedStatement.setInt(1, customerId);

      try (ResultSet resultSet = preparedStatement.executeQuery()) {
        if (!resultSet.next()) {
          // If the customer info for the specified customer ID doesn't exist, throw an exception
          throw new RuntimeException("Customer not found");
        }

        customerName = resultSet.getString("name");
      }
    }

    List<String> statementJsons = new ArrayList<>();
    int total = 0;

    // Retrieve the order statements for the order ID from the statements table
    try (PreparedStatement preparedStatementForStatements =
        connection.prepareStatement("SELECT * FROM sample.statements WHERE order_id = ?")) {
      preparedStatementForStatements.setString(1, orderId);

      try (ResultSet resultSetForStatements = preparedStatementForStatements.executeQuery()) {
        while (resultSetForStatements.next()) {
          int itemId = resultSetForStatements.getInt("item_id");

          // Retrieve the item data from the items table
          try (PreparedStatement preparedStatementForItems =
              connection.prepareStatement("SELECT * FROM sample.items WHERE item_id = ?")) {
            preparedStatementForItems.setInt(1, itemId);

            try (ResultSet resultSetForItems = preparedStatementForItems.executeQuery()) {
              if (!resultSetForItems.next()) {
                throw new RuntimeException("Item not found");
              }

              int price = resultSetForItems.getInt("price");
              int count = resultSetForStatements.getInt("count");

              // Make the statements JSON
              statementJsons.add(
                  String.format(
                      "{\"item_id\": %d, \"name\": \"%s\", \"price\": %d, \"count\": %d}",
                      itemId, resultSetForItems.getString("name"), price, count));

              // Calculate the total amount
              total += price * count;
            }
          }
        }
      }

      // Return the order info as a JSON format
      return String.format(
          "{\"order_id\": \"%s\",\"timestamp\": %d,\"customer_id\": %d,\"customer_name\": \"%s\",\"statement\": [%s],\"total\": %d}",
          orderId, timestamp, customerId, customerName, String.join(",", statementJsons), total);
    }
  }

  public String getOrderByOrderId(String orderId) throws SQLException {
    try (Connection connection = getConnection()) {
      try {
        // Get an order JSON for the specified order ID
        String orderJson = getOrderJson(connection, orderId);

        // Commit the transaction
        connection.commit();

        // Return the order info as a JSON format
        return String.format("{\"order\": %s}", orderJson);
      } catch (Exception e) {
        connection.rollback();
        throw e;
      }
    }
  }

  public String getOrdersByCustomerId(int customerId) throws SQLException {
    try (Connection connection = getConnection()) {
      // Retrieve the order info for the customer ID from the orders table
      try (PreparedStatement preparedStatement =
          connection.prepareStatement("SELECT * FROM sample.orders WHERE customer_id = ?")) {
        preparedStatement.setInt(1, customerId);

        List<String> orderJsons = new ArrayList<>();
        try (ResultSet resultSet = preparedStatement.executeQuery()) {
          // Make order JSONs for the orders of the customer
          while (resultSet.next()) {
            orderJsons.add(getOrderJson(connection, resultSet.getString("order_id")));
          }
        }

        // Commit the transaction
        connection.commit();

        // Return the order info as a JSON format
        return String.format("{\"order\": [%s]}", String.join(",", orderJsons));
      } catch (Exception e) {
        connection.rollback();
        throw e;
      }
    }
  }

  public void repayment(int customerId, int amount) throws SQLException {
    try (Connection connection = getConnection()) {
      try {
        // Retrieve the customer info for the specified customer ID from the customers table
        int updatedCreditTotal;
        try (PreparedStatement preparedStatement =
            connection.prepareStatement("SELECT * FROM sample.customers WHERE customer_id = ?")) {
          preparedStatement.setInt(1, customerId);

          try (ResultSet resultSet = preparedStatement.executeQuery()) {
            if (!resultSet.next()) {
              // If the customer info for the specified customer ID doesn't exist, throw an
              // exception
              throw new RuntimeException("Customer not found");
            }

            updatedCreditTotal = resultSet.getInt("credit_total") - amount;

            // Check if over repayment or not
            if (updatedCreditTotal < 0) {
              throw new RuntimeException("Over repayment");
            }
          }
        }

        // Reduce credit_total for the customer
        try (PreparedStatement preparedStatement =
            connection.prepareStatement(
                "UPDATE sample.customers SET credit_total = ? WHERE customer_id = ?")) {
          preparedStatement.setInt(1, updatedCreditTotal);
          preparedStatement.setInt(2, customerId);
          preparedStatement.executeUpdate();
        }

        // Commit the transaction
        connection.commit();
      } catch (Exception e) {
        connection.rollback();
        throw e;
      }
    }
  }

  private Connection getConnection() throws SQLException {
    Connection connection = DriverManager.getConnection("jdbc:scalardb:scalardb-sql.properties");
    connection.setAutoCommit(false);
    return connection;
  }
}

File: dotnet/scalardb-cluster-sample/Sample.cs
using System.Diagnostics;
using System.Text.Json.Nodes;
using Microsoft.Extensions.Logging;
using ScalarDB.Client;
using ScalarDB.Client.Builders;
using ScalarDB.Client.Builders.Admin;
using ScalarDB.Client.Core;
using ScalarDB.Client.Exceptions;

namespace ScalarDbClusterSample;

public class Sample: IDisposable
{
    private readonly TransactionFactory _factory;
    private readonly IDistributedTransactionManager _manager;

    public Sample()
    {
        var loggerFactory = LoggerFactory.Create(builder =>
        {
            builder.SetMinimumLevel(LogLevel.Warning);
            builder.AddSimpleConsole(options => { options.TimestampFormat = "HH:mm:ss "; });
        });
        
        _factory = TransactionFactory.Create("scalardb-options.json", loggerFactory);
        _manager = _factory.GetTransactionManager();
    }

    public async Task CreateTables()
    {
        using var admin = _factory.GetTransactionAdmin();
        
        var customersTableMetadata
            = new TableMetadataBuilder()
              .AddPartitionKey("customer_id", DataType.Int)
              .AddColumn("name", DataType.Text)
              .AddColumn("credit_limit", DataType.Int)
              .AddColumn("credit_total", DataType.Int)
              .Build();

        var ordersTableMetadata
            = new TableMetadataBuilder()
              .AddSecondaryIndex("order_id", DataType.Text)
              .AddPartitionKey("customer_id", DataType.Int)
              .AddClusteringKey("timestamp", DataType.Bigint)
              .Build();
        
        var statementsTableMetadata
            = new TableMetadataBuilder()
                .AddPartitionKey("order_id", DataType.Text)
                .AddClusteringKey("item_id", DataType.Int)
                .AddColumn("count", DataType.Int)
                .Build();
        
        var itemsTableMetadata
            = new TableMetadataBuilder()
                .AddPartitionKey("item_id", DataType.Int)
                .AddColumn("name", DataType.Text)
                .AddColumn("price", DataType.Int)
                .Build();
        
        await admin.CreateCoordinatorTablesAsync(true);
        await admin.CreateNamespaceAsync("sample", true);
        
        await admin.CreateTableAsync("sample", "customers", customersTableMetadata, true);
        await admin.CreateTableAsync("sample", "orders", ordersTableMetadata, true);
        await admin.CreateTableAsync("sample", "statements", statementsTableMetadata, true);
        await admin.CreateTableAsync("sample", "items", itemsTableMetadata, true);
    }

    public async Task LoadInitialData()
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            await LoadCustomerIfNotExists(1, "Yamada Taro", 10000, 0);
            await LoadCustomerIfNotExists(2, "Yamada Hanako", 10000, 0);
            await LoadCustomerIfNotExists(3, "Suzuki Ichiro", 10000, 0);
            await LoadItemIfNotExists(1, "Apple", 1000);
            await LoadItemIfNotExists(2, "Orange", 2000);
            await LoadItemIfNotExists(3, "Grape", 2500);
            await LoadItemIfNotExists(4, "Mango", 5000);
            await LoadItemIfNotExists(5, "Melon", 3000);
            
            await transaction.CommitAsync();
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
        
        return;

        async Task LoadCustomerIfNotExists(int customerId, string name,
                                           int creditLimit, int creditTotal)
        {
            var customer =
                await transaction.GetAsync(
                    new GetBuilder()
                        .SetNamespaceName("sample")
                        .SetTableName("customers")
                        .AddPartitionKey("customer_id", customerId)
                        .Build()
                );

            if (customer is not null)
                return;

            await transaction.InsertAsync(
                new InsertBuilder()
                    .SetNamespaceName("sample")
                    .SetTableName("customers")
                    .AddPartitionKey("customer_id", customerId)
                    .AddColumn("name", name)
                    .AddColumn("credit_limit", creditLimit)
                    .AddColumn("credit_total", creditTotal)
                    .Build()
            );
        }

        async Task LoadItemIfNotExists(int itemId, string name, int price)
        {
            var item =
                await transaction.GetAsync(
                    new GetBuilder()
                        .SetNamespaceName("sample")
                        .SetTableName("items")
                        .AddPartitionKey("item_id", itemId)
                        .Build()
                );

            if (item is not null)
                return;

            await transaction.InsertAsync(
                new InsertBuilder()
                    .SetNamespaceName("sample")
                    .SetTableName("items")
                    .AddPartitionKey("item_id", itemId)
                    .AddColumn("name", name)
                    .AddColumn("price", price)
                    .Build()
            );
        }
    }

    public async Task<JsonObject> GetCustomerInfo(int customerId)
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            // Retrieve the customer info for the specified customer ID from the customers table
            var customer =
                await transaction.GetAsync(
                    new GetBuilder()
                        .SetNamespaceName("sample")
                        .SetTableName("customers")
                        .AddPartitionKey("customer_id", customerId)
                        .Build()
                );

            if (customer is null)
            {
                // If the customer info the specified customer ID doesn't exist, throw an exception
                throw new Exception($"Customer not found (id: {customerId})");
            }

            // Commit the transaction (even when the transaction is read-only, we need to commit)
            await transaction.CommitAsync();

            return new JsonObject
                   {
                       { "id", customerId },
                       { "name", customer.GetValue<string>("name") },
                       { "credit_limit", customer.GetValue<int>("credit_limit") },
                       { "credit_total", customer.GetValue<int>("credit_total") }
                   };
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
    }

    public async Task<JsonObject> PlaceOrder(int customerId,
                                             IReadOnlyDictionary<int, int> itemCounts)
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            var orderId = Guid.NewGuid().ToString();

            // Insert the order info into the orders table
            await transaction.InsertAsync(
                new InsertBuilder()
                    .SetNamespaceName("sample")
                    .SetTableName("orders")
                    .AddPartitionKey("customer_id", customerId)
                    .AddClusteringKey("timestamp", DateTimeOffset.UtcNow.ToUnixTimeMilliseconds())
                    .AddColumn("order_id", orderId)
                    .Build()
            );

            var amount = 0;
            foreach (var (itemId, count) in itemCounts)
            {
                // Insert the order statement into the statements table
                await transaction.InsertAsync(
                    new InsertBuilder()
                        .SetNamespaceName("sample")
                        .SetTableName("statements")
                        .AddPartitionKey("order_id", orderId)
                        .AddClusteringKey("item_id", itemId)
                        .AddColumn("count", count)
                        .Build()
                );

                // Retrieve the item info from the items table
                var item =
                    await transaction.GetAsync(
                        new GetBuilder()
                            .SetNamespaceName("sample")
                            .SetTableName("items")
                            .AddPartitionKey("item_id", itemId)
                            .Build()
                    );
                if (item is null)
                    throw new Exception($"Item not found (id: {itemId})");

                // Calculate the total amount
                amount += item.GetValue<int>("price") * count;
            }

            // Check if the credit total exceeds the credit limit after payment
            var customer =
                await transaction.GetAsync(
                    new GetBuilder()
                        .SetNamespaceName("sample")
                        .SetTableName("customers")
                        .AddPartitionKey("customer_id", customerId)
                        .Build()
                );
            if (customer is null)
                throw new Exception($"Customer not found (id: {customerId})");

            var creditLimit = customer.GetValue<int>("credit_limit");
            var creditTotal = customer.GetValue<int>("credit_total");
            var newCreditTotal = creditTotal + amount;
            if (newCreditTotal > creditLimit)
                throw new Exception($"Credit limit exceeded ({newCreditTotal} > {creditLimit})");

            // Update credit_total for the customer
            await transaction.UpdateAsync(
                new UpdateBuilder()
                    .SetNamespaceName("sample")
                    .SetTableName("customers")
                    .AddPartitionKey("customer_id", customerId)
                    .AddColumn("credit_total", newCreditTotal)
                    .Build()
            );

            // Commit the transaction
            await transaction.CommitAsync();

            return new JsonObject
                   {
                       { "order_id", orderId }
                   };
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
    }

    public async Task<JsonObject> GetOrderByOrderId(string orderId)
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            // Get an order JSON for the specified order ID
            var orderJson = await getOrder(transaction, orderId);

            // Commit the transaction (even when the transaction is read-only, we need to commit)
            await transaction.CommitAsync();

            return new JsonObject
                   {
                       { "order", orderJson }
                   };
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
    }

    public async Task<JsonObject> GetOrdersByCustomerId(int customerId)
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            // Retrieve the order info for the customer ID from the orders table
            var orders =
                await transaction.ScanAsync(
                    new ScanBuilder()
                        .SetNamespaceName("sample")
                        .SetTableName("orders")
                        .AddPartitionKey("customer_id", customerId)
                        .Build()
                );

            // Make order JSONs for the orders of the customer
            var orderJsons = new JsonArray();
            foreach (var order in orders)
                orderJsons.Add(await getOrder(transaction, order.GetValue<string>("order_id")));

            // Commit the transaction (even when the transaction is read-only, we need to commit)
            await transaction.CommitAsync();

            return new JsonObject
                   {
                       { "orders", orderJsons }
                   };
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
    }

    public async Task Repayment(int customerId, int amount)
    {
        var transaction = await _manager.BeginAsync();
        try
        {
            // Retrieve the customer info for the specified customer ID from the customers table
            var customer =
                await transaction.GetAsync(
                    new GetBuilder()
                        .SetNamespaceName("sample")
                        .SetTableName("customers")
                        .AddPartitionKey("customer_id", customerId)
                        .Build()
                );
            if (customer is null)
                throw new Exception($"Customer not found (id: {customerId})");

            var updatedCreditTotal = customer.GetValue<int>("credit_total") - amount;

            // Check if over repayment or not
            if (updatedCreditTotal < 0)
                throw new Exception($"Over-repayment ({updatedCreditTotal})");

            // Reduce credit_total for the customer
            await transaction.UpdateAsync(
                new UpdateBuilder()
                    .SetNamespaceName("sample")
                    .SetTableName("customers")
                    .AddPartitionKey("customer_id", customerId)
                    .AddColumn("credit_total", updatedCreditTotal)
                    .Build()
            );

            // Commit the transaction
            await transaction.CommitAsync();
        }
        catch (TransactionException)
        {
            // If an error occurs, rollback the transaction
            await transaction.RollbackAsync();
            throw;
        }
    }

    private static async Task<JsonObject> getOrder(IDistributedTransaction transaction, string orderId)
    {
        // Retrieve the order info for the order ID from the orders table
        var order =
            await transaction.GetAsync(
                new GetBuilder()
                    .SetNamespaceName("sample")
                    .SetTableName("orders")
                    .SetGetType(GetOperationType.GetWithIndex)
                    .AddPartitionKey("order_id", orderId)
                    .Build()
            );
        if (order is null)
            throw new Exception($"Order not found (id: {orderId})");

        var customerId = order.GetValue<int>("customer_id");

        // Retrieve the customer info for the specified customer ID from the customers table
        var customer =
            await transaction.GetAsync(
                new GetBuilder()
                    .SetNamespaceName("sample")
                    .SetTableName("customers")
                    .AddPartitionKey("customer_id", customerId)
                    .Build()
            );
        Debug.Assert(customer is not null);

        // Retrieve the order statements for the order ID from the statements table
        var statements =
            await transaction.ScanAsync(
                new ScanBuilder()
                    .SetNamespaceName("sample")
                    .SetTableName("statements")
                    .AddPartitionKey("order_id", orderId)
                    .Build()
            );

        // Make the statements JSONs
        var total = 0;
        var statementJsons = new JsonArray();

        foreach (var statement in statements)
        {
            var itemId = statement.GetValue<int>("item_id");

            // Retrieve the item data from the items table
            var item =
                await transaction.GetAsync(
                    new GetBuilder()
                        .SetNamespaceName("sample")
                        .SetTableName("items")
                        .AddPartitionKey("item_id", itemId)
                        .Build()
                );
            if (item is null)
                throw new Exception($"Item not found (id: {itemId})");

            var price = item.GetValue<int>("price");
            var count = statement.GetValue<int>("count");
            var totalForStatement = price * count;

            statementJsons.Add(
                new JsonObject
                {
                    { "item_id", itemId },
                    { "item_name", item.GetValue<string>("name") },
                    { "price", price },
                    { "count", count },
                    { "total", totalForStatement }
                }
            );

            total += totalForStatement;
        }
        
        return new JsonObject
               {
                   { "order_id", orderId },
                   { "timestamp", order.GetValue<long>("timestamp") },
                   { "customer_id", customerId },
                   { "customer_name", customer.GetValue<string>("name") },
                   { "statements", statementJsons },
                   { "total", total }
               };
    }

    public void Dispose()
        => _manager.Dispose();
}
